### DISCUSSION TITLE
Remix on Vercel - Error Route handling

### DISCUSSION THREAD

**Discussion by sergical** [vercel/community#4596] [Help] [answered]
### Summary

The `ErrorBoundary` route gets built when you have any Edge functions in your app as an Edge function.

`main`: https://github.com/sergical/remix-playground/blob/main/app/routes/edge.tsx

<img width="1219" alt="Screenshot 2023-10-24 at 5 05 58 PM" src="https://github.com/vercel/community/assets/3760543/4388dbe1-9e2e-4786-a2f2-ac798516f553">

`not-edge`: https://github.com/sergical/remix-playground/blob/not-edge/app/routes/edge.tsx

<img width="1209" alt="Screenshot 2023-10-24 at 5 12 39 PM" src="https://github.com/vercel/community/assets/3760543/0322f86d-00c6-4df8-8d63-53d146bed271">

The actual `Edge` Error boundary works:

https://remix-playground.vercel.app/edge?error=error

<img width="1680" alt="Screenshot 2023-10-24 at 5 14 30 PM" src="https://github.com/vercel/community/assets/3760543/adf4bdfe-15dc-4e6e-8863-f265bf3cdb77">

However the root level error boundary crashes: 

<img width="1680" alt="Screenshot 2023-10-24 at 5 14 21 PM" src="https://github.com/vercel/community/assets/3760543/707d674c-bf70-4cb9-b14c-d405a645c17a">

And it doesn't crash on the `not-edge` branch:

https://remix-playground-git-not-edge-sergiy.vercel.app/adsf

<img width="1680" alt="Screenshot 2023-10-24 at 5 26 39 PM" src="https://github.com/vercel/community/assets/3760543/343a01fe-9dd9-4363-b231-182e12098d01">

I understand that root `ErrorBoundary` is what gets fed into the `404` Vercel function. But I was hoping we could have some flexibility on it not being `Edge` when it doesn't need to be. For example, each `edge` file has it's own ErrorBoundary so the root 404 can be built as Serverless. Otherwise all the default error handling that remix provides does not work and it becomes a bad DX and UX :sad

### Example

https://remix-playground.vercel.app/

### Steps to Reproduce

1. Clone https://github.com/sergical/remix-playground
2. Deploy to Vercel
3. Notice 404 being built as Edge
4. Create separate deployment similar to https://github.com/sergical/remix-playground/blob/main/app/routes/edge.tsx#L4
5. Notice 404 being built as Serverless

**Comment by TooTallNate** [comment, marked as answer]
Thanks for the repro instructions.

The error during Edge rendering seems to come from `@sentry/remix`:

```
TypeError: (void 0) is not a function
    at (build/build-edge-eyJydW50aW1lIjoiZWRnZSJ9.js:18:0)
    at (node_modules/@remix-run/server-runtime/dist/esm/server.js:39:0)
    at (node_modules/@remix-run/server-runtime/dist/esm/server.js:152:0)
    at (node_modules/@remix-run/server-runtime/dist/esm/server.js:149:0)
    at (node_modules/@remix-run/server-runtime/dist/esm/server.js:59:0)
```

Which is this line:

```js
  Sentry.captureRemixServerException(error, "remix.server", request);
```

When Vercel bundles Edge functions, it prefers the "browser" field in the `package.json`, which ends up being incompatible with with the server-side API of that package.

To be compatible with Vercel Edge functions, I would expect `@sentry/remix` to define an "exports" field in `package.json` with an "edge-light" field pointing to a compatible entrypoint file, but that is not the case today (as of `@sentry/remix@7.73.0`).

[This issue](https://github.com/getsentry/sentry-javascript/issues/5610) seems somewhat related. They mention about WinterCG compatibility, which should make Edge functions work if that were implemented. However, for now, it seems like this `@sentry/remix` package is only compatible with Node.js.

**Comment by sergical** [comment]
Thank you for the explanation, really helpful context! 

Can confirm https://github.com/sergical/remix-playground/pull/4 - https://remix-playground-git-no-sentry-server-sergiy.vercel.app/asdf renders ErrorBoundary as expected

Understanding the edge function failure, however, could you share a bit more insight into how Vercel handles deciding between Edge 404 vs SSR - is it looking for any Edge functions in the build and decides? There's no way to control it?

FWIW Sentry has an [alpha edge support adapter](https://www.npmjs.com/package/@sentry/vercel-edge) which I can try

**Reply by TooTallNate** [reply]
The current logic is that, if there's any Edge route, then the 404 page will be Edge. Given your use case, I think it would make sense to re-think that, and instead respect the root level config. We'll be looking into that.

**Reply by TooTallNate** [reply]
Ok so I finally was able to look into this. The answer is that you should define a catch-all splat route in your Remix app (i.e. `routes/$.tsx`), with the contents something like:

```typescript
import { json, useLocation } from "@remix-run/react";

export const loader = () => {
  return json(null, { status: 404 });
};

export default function Custom404() {
  const location = useLocation();
  return <h1>Custom 404: {location.pathname}</h1>;
}
```

You can do any additional logic here like reporting to Sentry. Let me know if that works for you!
